File: finops-micro/backend/src/finops_api/schemas/finops.py
from __future__ import annotations

from datetime import date

from pydantic import BaseModel, Field


class PeakDay(BaseModel):
    date: date
    amount: float


class SummaryV2Response(BaseModel):
    totalWeek: float
    deltaWeek: float
    avgDaily: float
    peakDay: PeakDay
    monthTotal: float
    yearTotal: float
    budgetMonth: float | None = None
    budgetYear: float | None = None
    usdRate: float | None = None


class DailyItem(BaseModel):
    date: date
    total: float
    byService: dict[str, float] | None = None


class RankedItemV2(BaseModel):
    serviceName: str | None = None
    linkedAccount: str | None = None
    total: float
    sharePct: float
    delta: float
    deltaPct: float


class FiltersV2Response(BaseModel):
    services: list[str] = Field(default_factory=list)
    accounts: list[str] = Field(default_factory=list)


class TenantOption(BaseModel):
    tenantKey: str
    tenantName: str
    cloud: str


class AiInsightRequest(BaseModel):
    cloud: str
    tenant_key: str | None = None
    from_: date = Field(alias="from")
    to: date
    currency: str
    question: str
    filters: dict[str, list[str]] | None = None


class AiInsightResponse(BaseModel):
    answerMarkdown: str
    highlights: list[str]
    suggestedActions: list[str]


class AnalyticsInsightRequest(BaseModel):
    cloud: str
    tenant_key: str | None = None
    from_: date = Field(alias="from")
    to: date
    currency: str
    topN: int = 10
    services: list[str] | None = None
    accounts: list[str] | None = None


class AnalyticsInsightEvidence(BaseModel):
    topServices: list[str] = Field(default_factory=list)
    topAccounts: list[str] = Field(default_factory=list)
    peakDay: date | None = None
    totalPeriod: float
    deltaPeriodPct: float


class AnalyticsInsightAction(BaseModel):
    title: str
    owner: str
    priority: str
    rationale: str


class AnalyticsInsightResponse(BaseModel):
    mode: str
    summary: str
    drivers: list[str] = Field(default_factory=list)
    risks: list[str] = Field(default_factory=list)
    actions: list[AnalyticsInsightAction] = Field(default_factory=list)
    suggestedQuestions: list[str] = Field(default_factory=list)
    evidence: AnalyticsInsightEvidence


class CostExplorerSnapshotKpi(BaseModel):
    label: str
    value: float
    sharePct: float
    deltaPct: float


class CostExplorerSnapshotResponse(BaseModel):
    totalPeriod: float
    deltaPeriodPct: float
    top1SharePct: float
    top3SharePct: float
    peakDay: PeakDay
    largestService: CostExplorerSnapshotKpi | None = None
    largestAccount: CostExplorerSnapshotKpi | None = None


class CostExplorerBreakdownItem(BaseModel):
    key: str
    label: str
    total: float
    sharePct: float
    delta: float
    deltaPct: float
    contributionPct: float


class CostExplorerTrendItem(BaseModel):
    date: date
    total: float
    selected: float
    others: float


class CostExplorerInsightRequest(BaseModel):
    cloud: str
    tenant_key: str | None = None
    from_: date = Field(alias="from")
    to: date
    currency: str
    topN: int = 10
    groupBy: str = "service"
    selectedItem: str | None = None
    services: list[str] | None = None
    accounts: list[str] | None = None


class CostExplorerInsightEvidence(BaseModel):
    groupBy: str
    selectedItem: str | None = None
    totalPeriod: float
    deltaPeriodPct: float
    peakDay: date | None = None
    topBreakdown: list[str] = Field(default_factory=list)
    topServices: list[str] = Field(default_factory=list)
    topAccounts: list[str] = Field(default_factory=list)


class CostExplorerInsightAction(BaseModel):
    title: str
    owner: str
    priority: str
    rationale: str


class CostExplorerNextDrilldown(BaseModel):
    dimension: str
    value: str
    reason: str


class CostExplorerInsightResponse(BaseModel):
    mode: str
    summary: str
    drivers: list[str] = Field(default_factory=list)
    risks: list[str] = Field(default_factory=list)
    actions: list[CostExplorerInsightAction] = Field(default_factory=list)
    suggestedQuestions: list[str] = Field(default_factory=list)
    nextDrilldowns: list[CostExplorerNextDrilldown] = Field(default_factory=list)
    evidence: CostExplorerInsightEvidence


class ReingestRequest(BaseModel):
    cloud: str
    tenant_key: str | None = None
    from_: date = Field(alias="from")
    to: date


class ReingestResult(BaseModel):
    provider: str
    rows_received: int
    rows_written: int


class ReingestResponse(BaseModel):
    results: list[ReingestResult]

File: finops-micro/frontend/src/lib/schemas/finops.ts
import { z } from "zod";

export const summarySchema = z.object({
  totalWeek: z.number(),
  deltaWeek: z.number(),
  avgDaily: z.number(),
  peakDay: z.object({ date: z.string(), amount: z.number() }),
  monthTotal: z.number(),
  yearTotal: z.number(),
  budgetMonth: z.number().nullish(),
  budgetYear: z.number().nullish(),
  usdRate: z.number().nullish(),
});

export const dailyItemSchema = z.object({
  date: z.string(),
  total: z.number(),
  byService: z.record(z.number()).optional(),
});

export const topItemSchema = z.object({
  serviceName: z.string().nullish(),
  linkedAccount: z.string().nullish(),
  total: z.number(),
  sharePct: z.number(),
  delta: z.number(),
  deltaPct: z.number(),
});

export const filtersSchema = z.object({
  services: z.array(z.string()),
  accounts: z.array(z.string()),
});

export const aiInsightSchema = z.object({
  answerMarkdown: z.string(),
  highlights: z.array(z.string()),
  suggestedActions: z.array(z.string()),
});

export const analyticsInsightSchema = z.object({
  mode: z.string(),
  summary: z.string(),
  drivers: z.array(z.string()),
  risks: z.array(z.string()),
  actions: z.array(
    z.object({
      title: z.string(),
      owner: z.string(),
      priority: z.string(),
      rationale: z.string(),
    }),
  ),
  suggestedQuestions: z.array(z.string()),
  evidence: z.object({
    topServices: z.array(z.string()),
    topAccounts: z.array(z.string()),
    peakDay: z.string().nullish(),
    totalPeriod: z.number(),
    deltaPeriodPct: z.number(),
  }),
});

export const costExplorerSnapshotSchema = z.object({
  totalPeriod: z.number(),
  deltaPeriodPct: z.number(),
  top1SharePct: z.number(),
  top3SharePct: z.number(),
  peakDay: z.object({ date: z.string(), amount: z.number() }),
  largestService: z
    .object({
      label: z.string(),
      value: z.number(),
      sharePct: z.number(),
      deltaPct: z.number(),
    })
    .nullish(),
  largestAccount: z
    .object({
      label: z.string(),
      value: z.number(),
      sharePct: z.number(),
      deltaPct: z.number(),
    })
    .nullish(),
});

export const costExplorerBreakdownSchema = z.array(
  z.object({
    key: z.string(),
    label: z.string(),
    total: z.number(),
    sharePct: z.number(),
    delta: z.number(),
    deltaPct: z.number(),
    contributionPct: z.number(),
  }),
);

export const costExplorerTrendSchema = z.array(
  z.object({
    date: z.string(),
    total: z.number(),
    selected: z.number(),
    others: z.number(),
  }),
);

export const costExplorerInsightSchema = z.object({
  mode: z.string(),
  summary: z.string(),
  drivers: z.array(z.string()),
  risks: z.array(z.string()),
  actions: z.array(
    z.object({
      title: z.string(),
      owner: z.string(),
      priority: z.string(),
      rationale: z.string(),
    }),
  ),
  suggestedQuestions: z.array(z.string()),
  nextDrilldowns: z.array(
    z.object({
      dimension: z.string(),
      value: z.string(),
      reason: z.string(),
    }),
  ),
  evidence: z.object({
    groupBy: z.string(),
    selectedItem: z.string().nullish(),
    totalPeriod: z.number(),
    deltaPeriodPct: z.number(),
    peakDay: z.string().nullish(),
    topBreakdown: z.array(z.string()),
    topServices: z.array(z.string()),
    topAccounts: z.array(z.string()),
  }),
});

export const dailySchema = z.array(dailyItemSchema);
export const topServicesSchema = z.array(
  topItemSchema.transform((item) => ({
    ...item,
    serviceName: item.serviceName ?? "N/A",
  })),
);
export const topAccountsSchema = z.array(
  topItemSchema.transform((item) => ({
    ...item,
    linkedAccount: item.linkedAccount ?? "N/A",
  })),
);

export type SummaryResponse = z.infer<typeof summarySchema>;
export type DailyResponse = z.infer<typeof dailySchema>;
export type TopServicesResponse = z.infer<typeof topServicesSchema>;
export type TopAccountsResponse = z.infer<typeof topAccountsSchema>;
export type FiltersResponse = z.infer<typeof filtersSchema>;
export type AiInsightResponse = z.infer<typeof aiInsightSchema>;
export type AnalyticsInsightResponse = z.infer<typeof analyticsInsightSchema>;
export type CostExplorerSnapshotResponse = z.infer<typeof costExplorerSnapshotSchema>;
export type CostExplorerBreakdownResponse = z.infer<typeof costExplorerBreakdownSchema>;
export type CostExplorerTrendResponse = z.infer<typeof costExplorerTrendSchema>;
export type CostExplorerInsightResponse = z.infer<typeof costExplorerInsightSchema>;
